test/misc/misc.js

```diff
@@ -3,6 +3,16 @@ const rollup = require('../../dist/rollup');
 const { loader } = require('../utils.js');
 
 describe('misc', () => {
+	it('throw modification of options.acorn', () => {
+		return rollup
+			.rollup({
+				input: 'input',
+				plugins: [ loader({ input: `export default 0;` }) ],
+				acorn: Object.freeze({}),
+				experimentalTopLevelAwait: true,
+			});
+	});
+
```

Comment:
How about making this test a little deeper by doing it like this:

```js
	it('throw modification of options.acorn', () => {
		const options = {
				input: 'input',
				plugins: [ loader({ input: `export default 0;` }) ],
				acorn: {},
				experimentalTopLevelAwait: true,
			};
		return rollup
			.rollup(options).then(() => assert.deepStrictEqual(options, {
				input: 'input',
				plugins: [ loader({ input: `export default 0;` }) ],
				acorn: {},
				experimentalTopLevelAwait: true,
			});
	});
```

Or does this actually fail? That way, we would also check that other options are protected and we can easily extend the test to check more stuff.

Reply:
Ok, this might have issues with the plugins, but I think it is worth a try.

Reply:
The plugin returned by `loader` has no `name` field, and rollup will set a name for it. I can't remove all plugin.name modification in this PR, that's a big project. The easy way is add name in loader utils directly. Of cause, the easiest way is ignore plugin related modification.

But what I wonder is, since rollup is run in strict mode, just freeze the options, then if there is any modification to the options, an error will be thrown. Do you think there is any defect in this way?

Reply:
And assert.deepStrictEqual also can't promise no modification, if rollup set options' property value to another object (same structure), but not same object any more. Then if user had an refference to the original value object, and want to modify it, unexpected things will happen.

```js
const acorn = { ecmaVersion: 6 };
const options = { acorn };
await rollup(options);// suppose rollup set options.acorn = Object.assign({}, options.acorn);
acorn.ecmaVersion = 7;
await rollup(options);// works unexpected
```

Reply:
> just freeze the options, then if there is any modification to the options

You a re right, the thing with `Object.freeze` is that it only does a shallow freeze but not a "deep" freeze. Which is ok for this specific issue but would not work for checking the whole options object. But as you said since we are mutating plugins anyway, let's stick with what we have.

Reply:
Hi, you maybe miss my second message ;)

> And assert.deepStrictEqual also can't promise no modification, if rollup set options' property value to another object (same structure), but not same object any more. Then if user had an refference to the original value object, and want to modify it, unexpected things will happen.
> 
> ```js
> const acorn = { ecmaVersion: 6 };
> const options = { acorn };
> await rollup(options);// suppose rollup set options.acorn = Object.assign({}, options.acorn);
> acorn.ecmaVersion = 7;
> await rollup(options);// works unexpected
> ```

`assert.deepStrictEqual` won't be perfect in modification checking.

And I already add `Object.freeze` to all non-primitive value to achieve the purpose that checking the whole options object in the latest changes.

And, rollup do modification to the options currenctly infact (add plugin.name and delete options.cache), we can't check them before we adapt all of them. I added two comments in the test too, so as to check the in the future.

Reply:
I read it and agreed, sorry, there is a lot going on here.